pkg/cmd/release-payload-controller/payload_creation_controller.go

```diff
@@ -0,0 +1,210 @@
+		releasePayloadNamespace: releasePayloadNamespace,
+		releasePayloadLister:    releasePayloadInformer.Lister(),
+		releasePayloadClient:    releasePayloadClient,
+		eventRecorder:           eventRecorder.WithComponentSuffix("payload-creation-controller"),
+		queue:                   workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PayloadCreationController"),
+	}
+
+	c.cachesToSync = append(c.cachesToSync, releasePayloadInformer.Informer().HasSynced)
+
+	releasePayloadInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: c.Enqueue,
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			c.Enqueue(newObj)
+		},
+		DeleteFunc: c.Enqueue,
+	})
+
+	return c, nil
+}
+
+func (c *PayloadCreationController) Enqueue(obj interface{}) {
+	key, err := cache.MetaNamespaceKeyFunc(obj)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid queue key '%v': %v", obj, err))
+		return
+	}
+	c.queue.Add(key)
+}
+
+func (c *PayloadCreationController) Run(ctx context.Context) {
+	defer utilruntime.HandleCrash()
+
+	klog.Info("Starting Payload Creation Controller")
+	defer func() {
+		klog.Info("Shutting down Payload Creation Controller")
+		c.queue.ShutDown()
+		klog.Info("Payload Creation Controller shut down")
+	}()
+
+	if !cache.WaitForNamedCacheSync("PayloadCreationController", ctx.Done(), c.cachesToSync...) {
+		return
+	}
+
+	go func() {
+		wait.UntilWithContext(ctx, c.runWorker, time.Second)
+	}()
+
+	<-ctx.Done()
+}
+
+func (c *PayloadCreationController) runWorker(ctx context.Context) {
+	for c.processNextItem(ctx) {
+	}
+}
+
+func (c *PayloadCreationController) processNextItem(ctx context.Context) bool {
+	key, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(key)
+
+	err := c.sync(ctx, key.(string))
+
+	if err == nil {
+		c.queue.Forget(key)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %w", key, err))
+	c.queue.AddRateLimited(key)
+
+	return true
+}
+
+func (c *PayloadCreationController) sync(ctx context.Context, key string) error {
+	klog.V(4).Infof("Starting PayloadCreationController sync")
+	defer klog.V(4).Infof("PayloadCreationController sync done")
+
+	// Convert the namespace/name string into a distinct namespace and name
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("invalid resource key: %s", key))
+		return nil
+	}
+
+	klog.V(4).Infof("Processing ReleasePayload: '%s/%s' from workQueue", namespace, name)
+
+	// Get the ReleasePayload resource with this namespace/name
+	originalReleasePayload, err := c.releasePayloadLister.ReleasePayloads(namespace).Get(name)
+	// The ReleasePayload resource may no longer exist, in which case we stop processing.
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	// If the conditions are both in their respective terminal states, then there is nothing else to do...
```

Comment:
keep this here *and* add it as a filter above.

Reply:
Done